# Graph contrastive learning as a versatile foundation for advanced scRNA-seq data analysis
![model](https://github.com/zhangzh1328/scSimGCL/blob/main/scSimGCL.png)

## Requirements
- python : 3.8.17
- scanpy : 1.9.6
- sklearn : 1.2.2
- torch : 1.8.1
- torch-geometric : 2.2.0

## Datasets
- Shekhar mouse retinca cells : https://www.ncbi.nlm.nih.gov/geo/query/acc.cgi?acc=GSE81904
- Baron : https://www.ncbi.nlm.nih.gov/geo/query/acc.cgi?acc=GSE84133
- 10X PBMC : https://support.10xgenomics.com/single-cell-gene-expression/datasets/2.1.0/pbmc4k
- Camp : https://www.ncbi.nlm.nih.gov/geo/query/acc.cgi?acc=GSE81252
- Mouse bladder cells : https://figshare.com/s/865e694ad06d5857db4b
- Zeisel : https://www.ncbi.nlm.nih.gov/geo/query/acc.cgi?acc=GSE60361
- Tabula Sapiens : https://figshare.com/projects/Tabula_Sapiens/100973
- Chien : https://www.ncbi.nlm.nih.gov/geo/query/acc.cgi?acc=GSE247988

## Usage  
The Baron dataset is taken as an exmaple: 
```bash
python main.py --data_path './Baron.h5' --save_model_path './save_file' --n_clusters 14
```
```z``` can be obtained from the ```Model``` class, one of the outputs of the ```Model``` class, which is a representation generated by scSimGCL. The ```z``` is used to create an AnnData object with ```sc.AnnData(z)```, which is then used in general scRNA-seq data analysis pipelines (e.g., scanpy).

## Citation
Zhang et al. "Graph contrastive learning as a versatile foundation for advanced scRNA-seq data analysis." ***Briefings in Bioinformatics*** 25.6 (2024): bbae558.
